jssrc/client.ts

```diff
@@ -0,0 +1,420 @@
+
+    // Update indexes
+    if(a === "name") indexes.byName.insert(v, e);
+    if(a === "tag") indexes.byTag.insert(v, e);
+    if(a === "class") indexes.byClass.insert(v, e);
+    if(a === "style") indexes.byStyle.insert(v, e);
+
+
+    // Update active*
+    if(a === "children") {
+      if(activeChildren[v]) console.error(`Unable to handle child element ${v} parented to two parents (${activeChildren[v]}, ${e}). Overwriting.`);
+      activeChildren[v] = e;
+    }
+
+    if(a === "value") {
+      entitiesWithUpdatedValues[e] = true;
+    }
+  }
+
+  // Update value syncing
+  for(let e in entitiesWithUpdatedValues) {
+    let a = "value";
+    let entity = entities[e];
+    if(!entity[a]) {
+      sentInputValues[e] = [];
+    } else {
+      if(entity[a].constructor === Array) console.error("Unable to set 'value' multiple times on entity", e, entity[a]);
+      let sent = sentInputValues[e];
+      if(sent && sent[0] === entity[a]) {
+        let ix;
+        while((ix = dirty[e].indexOf(a)) !== -1) {
+          dirty[e].splice(ix, 1);
+        }
+        sent.shift();
+      } else {
+        sentInputValues[e] = [];
+      }
+    }
+  }
+  for(let indexName in indexes) {
+    let index:Index = indexes[indexName];
+    index.dispatchIfDirty();
+  }
+}
+
+let prerendering = false;
+var frameRequested = false;
+
+var socket = new WebSocket("ws://" + window.location.host +"/ws");
+socket.onmessage = function(msg) {
+  let data = JSON.parse(msg.data);
+  if(data.type == "result") {
+    handleDiff(state, data);
+    renderRecords(state);
+
+    let diffEntities = 0;
+    if(DEBUG) {
+      console.groupCollapsed(`Received Result +${data.insert.length}/-${data.remove.length} (∂Entities: ${diffEntities})`);
+      if(DEBUG === true || DEBUG === "diff") {
+        console.table(data.insert);
+        console.table(data.remove);
+      }
+      if(DEBUG === true || DEBUG === "state") {
+        // we clone here to keep the entities fresh when you want to thumb through them in the log later (since they are rendered lazily)
+        let copy = clone(state.entities);
+
+        console.log("Entities", copy);
+        console.log("Indexes", indexes);
+      }
+      console.groupEnd();
+    }
+
+    if(document.readyState === "complete") {
+      renderEditor();
+    } else if(!prerendering) {
+      prerendering = true;
+      document.addEventListener("DOMContentLoaded", function() {
+        renderEditor();
+      });
+    }
+
+  } else if(data.type == "error") {
+    console.error(data.message, data);
+  }
+}
+socket.onopen = function() {
+  console.log("Connected to eve server!");
+  onHashChange({});
+}
+socket.onclose = function() {
+  console.log("Disconnected from eve server!");
+}
+
+//---------------------------------------------------------
+// Bootstrapping interface
+//---------------------------------------------------------
+export var parseInfo = {blocks: [], lines: []};
+
+let updateEditorParse = debounce(handleEditorParse, 1);
```

Comment:
We want to call this after any change that effected entities tagged `token` or `block`, not just when new entities are added/removed from the index. We can do that by cross-referencing the `dirty` entity list with the `indexes.byTag.index[<"block" | "token">]` entity lists